### PR TITLE
Tests for pollset_set.

### PR DESCRIPTION
**Pull request by sreecha**
Note: I initially created https://github.com/grpc/grpc/pull/9009 for this but that did not do it the right way. 

This is a test for `pollset_set`. This is meant to be a test for pollset_set implementations in all polling engines - but currently it only works for epoll engine (the poll engine's pollset_set implementation currently fails this test and I will be modifying that soon)

### CODE REVIEW

**Comment by sreecha** [comment]
Adding `DO NOT MERGE` label since it contains changes to `build.yaml` file (and since we are holding off all merges that modify `build.yaml` for now)

**Comment by sreecha** [comment]
All test passed. The failure 'Based Full Build & Tests' is a known issue.

**Comment by sreecha** [comment]
ping. @ctiller , can you take a look at this ? (else, this might result in merge conflicts and I'll have to wait for all the tests to run :) )

File: test/core/iomgr/pollset_set_test.c [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+            "Skipping the test. The test is only relevant for 'epoll' "
+            "strategy. and the current strategy is: '%s'",
+            poll_strategy);
+  }
+
+  grpc_iomgr_shutdown(&exec_ctx);
+  grpc_exec_ctx_finish(&exec_ctx);
+  return 0;
+}
+#else /* defined(GRPC_LINUX_EPOLL) */
```
Comment (dgquintas) [inline test/core/iomgr/pollset_set_test.c]: seems to me all you can say about that else is that `GRPC_POSIX_SOCKET` wouldn't be defined. Does `!GRPC_POSIX_SOCKET` imply `GRPC_LINUX_EPOLL`?

File: test/core/iomgr/pollset_set_test.c [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+#include "src/core/lib/iomgr/port.h"
+
+/* This test only relevant on linux systems */
+#ifdef GRPC_POSIX_SOCKET
+#include "src/core/lib/iomgr/ev_posix.h"
+
+#include <errno.h>
```
Comment (dgquintas) [inline test/core/iomgr/pollset_set_test.c]: move standard headers to the top of the #includes list?

File: test/core/iomgr/pollset_set_test.c [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+#include "src/core/lib/iomgr/port.h"
+
+/* This test only relevant on linux systems */
+#ifdef GRPC_POSIX_SOCKET
+#include "src/core/lib/iomgr/ev_posix.h"
```
Comment (dgquintas) [inline test/core/iomgr/pollset_set_test.c]: move src/... include to the bottom, after the <grpc/...> ones.

File: test/core/iomgr/pollset_set_test.c [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+#include "src/core/lib/iomgr/iomgr.h"
+#include "test/core/util/test_config.h"
+
+/*******************************************************************************
+ * test_pollset_set
+ */
+
+typedef struct test_pollset_set { grpc_pollset_set *pss; } test_pollset_set;
+
+void init_test_pollset_sets(test_pollset_set pollset_sets[], int num_pss) {
```
Comment (dgquintas) [inline test/core/iomgr/pollset_set_test.c]: we don't use the `foo[]` syntax for arguments. Change to `*pollset_sets`?

File: test/core/iomgr/pollset_set_test.c [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+  int i;
+  for (i = 0; i < num_pss; i++) {
```
Comment (dgquintas) [inline test/core/iomgr/pollset_set_test.c]: why not `for( int i = ... )` ?

File: test/core/iomgr/pollset_set_test.c [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+    pollset_sets[i].pss = grpc_pollset_set_create();
+  }
+}
+
+void cleanup_test_pollset_sets(test_pollset_set pollset_sets[], int num_pss) {
+  int i;
+  for (i = 0; i < num_pss; i++) {
```
Comment (dgquintas) [inline test/core/iomgr/pollset_set_test.c]: why not `for( int i = ... )` ?

File: test/core/iomgr/pollset_set_test.c [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+ */
+
+typedef struct test_pollset {
+  grpc_pollset *ps;
+  gpr_mu *mu;
+} test_pollset;
+
+static void init_test_pollsets(test_pollset pollsets[], int num_pollsets) {
+  int i;
+  for (i = 0; i < num_pollsets; i++) {
```
Comment (dgquintas) [inline test/core/iomgr/pollset_set_test.c]: why not `for( int i = ... )` ?

File: test/core/iomgr/pollset_set_test.c [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+                            grpc_error *error) {
+  grpc_pollset_destroy(p);
+}
+
+static void cleanup_test_pollsets(grpc_exec_ctx *exec_ctx,
+                                  test_pollset pollsets[], int num_pollsets) {
+  grpc_closure destroyed;
+  int i;
+
+  for (i = 0; i < num_pollsets; i++) {
```
Comment (dgquintas) [inline test/core/iomgr/pollset_set_test.c]: why not `for( int i = ... )` ? 

Ok I'm done repeating this :) Same applies for all the rest. It's just a nit btw.

File: test/core/iomgr/pollset_set_test.c [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+   */
+  grpc_exec_ctx exec_ctx = GRPC_EXEC_CTX_INIT;
+  int i;
+  grpc_pollset_worker *worker;
+  gpr_timespec deadline;
+
+  test_fd tfds[10];
+  test_pollset pollsets[3];
+  test_pollset_set pollset_sets[2];
+  int num_fds = sizeof(tfds) / sizeof(tfds[0]);
```
Comment (dgquintas) [inline test/core/iomgr/pollset_set_test.c]: we have `GPR_ARRAY_SIZE` in include/grpc/support/useful.h for this.

Also, you can make all these `num_*` variables `const` for better readability.

File: test/core/iomgr/pollset_set_test.c [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+   *                    +---> FD1
+   */
+  grpc_exec_ctx exec_ctx = GRPC_EXEC_CTX_INIT;
+  grpc_pollset_worker *worker;
+  gpr_timespec deadline;
+
+  test_fd tfds[3];
+  test_pollset pollset;
+  test_pollset_set pollset_sets[2];
+  int num_fds = sizeof(tfds) / sizeof(tfds[0]);
```
Comment (dgquintas) [inline test/core/iomgr/pollset_set_test.c]: ditto wrt `GPR_ARRAY_SIZE` and `const`

File: test/core/iomgr/pollset_set_test.c [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+   *                   +---> FD2
+   */
+  grpc_exec_ctx exec_ctx = GRPC_EXEC_CTX_INIT;
+  grpc_pollset_worker *worker;
+  gpr_timespec deadline;
+
+  test_fd tfds[3];
+  test_pollset pollsets[2];
+  test_pollset_set pollset_set;
+  int num_fds = sizeof(tfds) / sizeof(tfds[0]);
```
Comment (dgquintas) [inline test/core/iomgr/pollset_set_test.c]: ditto wrt `GPR_ARRAY_SIZE` and `const`

File: test/core/iomgr/pollset_set_test.c [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+  grpc_exec_ctx_finish(&exec_ctx);
+}
+
+int main(int argc, char **argv) {
+  const char *poll_strategy = NULL;
+  grpc_exec_ctx exec_ctx = GRPC_EXEC_CTX_INIT;
+  grpc_test_init(argc, argv);
+  grpc_iomgr_init();
+
+  poll_strategy = grpc_get_poll_strategy_name();
```
Comment (dgquintas) [inline test/core/iomgr/pollset_set_test.c]: you could just do `const char *poll_strategy = grpc_get...`.

File: test/core/iomgr/pollset_set_test.c (hunk 3 above)
Comment (sreecha) [inline test/core/iomgr/pollset_set_test.c]: Done

**Comment by dgquintas** [comment]
Bunch of compilation issues:

```
test/core/iomgr/pollset_set_test.c:271:16: error: implicit declaration of function 'GRPC_TIMEOUT_MILLIS_TO_DEADLINE' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    deadline = GRPC_TIMEOUT_MILLIS_TO_DEADLINE(2);
```
https://grpc-testing.appspot.com/job/gRPC_pull_requests_asan_c/1652/console

**Comment by sreecha** [comment]
Fixed the compilation issues. The failures now are all known. This can be merged.

**Pull request merged**